### ISSUE TITLE
How to turn off inline math highlights?

### ISSUE THREAD

**Issue by nathanbraun** [vimwiki/vimwiki#150]
I'm taking some notes where I occasionally have two dollar signs ($) in a line.  Due to (I think) vimwiki's inline math feature, anything between these is highlighted and the dollar signs are hidden.  Is there anyway to turn this feature off?

Currently I'm getting around it by entering $$ for every $, which makes the dollar signs appear, but still highlights everything in the middle.

**Comment by EinfachToll** [comment]
This is a principal problem with the lightweight markup languages -- if you want to use the characters which determine the markup as actual characters, you have to use ugly workarounds. In this case, the only workaround is probably to use backticks to escape the dollar signs: ``text `$` text `$` more text``.

**Comment by volker-fr** [comment]
I found a solution for presentation in vim itself. If you transform it into html it might end up as an equation.

```
" the following to line don't work... that would be the best way
let g:vimwiki_char_eqin = '\#\#'
let g:vimwiki_rxEqIn = ''

" Overwrite how it is represented inside in vim
hi def link VimwikiEqIn Normal
```

**Comment by protist** [comment]
@volker-fr With that workaround, I still need to type in a double $. Is there no way to totally ignore the math part of vimwiki? I rarely use it anyway.

Also, as you say, the other problem is that html export is broken. (Without MathJax installed) I see the following for the double $.

    foo has $\(10, bar has \)$20

And with a single $, which is hence hidden in vim itself, the html looks like

    foo has \(10, bar has \)20

IMO having more than one `$` in a line is not so obscure. I wonder if @EinfachToll might consider changing the math markup to `$$` or something slightly less common?

**Comment by protist** [comment]
I was trying to write a list that included prices, so I revisited this issue. It's extremely frustrating!

One (unsatisfying) workaround is to use one of the [alternative dollar glyphs](https://en.wikipedia.org/wiki/Dollar_sign#Encoding) for literal dollar signs. For example, the CJK fullwidth form U+FF04 (＄). However, in my terminal, this glyph is two characters wide, thus throwing out the alignment for the line.

**Comment by n0tHappy** [comment]
> I'm taking some notes where I occasionally have two dollar signs ($) in a line. Due to (I think) vimwiki's inline math feature, anything between these is highlighted and the dollar signs are hidden. Is there anyway to turn this feature off?
> 
> Currently I'm getting around it by entering $$ for every $, which makes the dollar signs appear, but still highlights everything in the middle.

hi,man. I find a way to solve this problem. we can modify the source code of vimwiki.
we can find in vimwiki/syntax/vimwiki_default.vim:
```
" text: $ equation_inline $
let s:default_syntax.rxEqIn = '\$[^$`]\+\$'
let s:default_syntax.char_eqin = '\$"'
```
and we change to:
```
let s:default_syntax.rxEqIn = '\✹[^✹`]\+\✹'
let s:default_syntax.char_eqin = '\✹"'
```
we can use a uncommon character to replace '$'.

**Comment by protist** [comment]
Thanks @n0tHappy! Good find! That works pretty well, although mine was slightly different.

Firstly, my original file and the [current version](https://github.com/vimwiki/vimwiki/blob/master/syntax/vimwiki_default.vim#L16) don't have a `"` near the end of the second line. Secondly, you don't need to escape `✹` the same way you escape `$`, so for it to replace the literal `$` in function, I used the following:
```
let s:default_syntax.rxEqIn = '✹[^✹`]\+✹'
let s:default_syntax.char_eqin = '✹'
```

I played around with having it in `~/.vimrc`, but that didn't really work (probably unsurprising).

I also tried using a more user-friendly glyph. I thought perhaps two characters would be nice, e.g. `$m`, but I couldn't work out how to write the regex. I was thinking negative lookahead, but couldn't fathom how to incorporate this. Something like `\$m\(.*$m\)\@!\$m`, but that didn't work, presumably because the lookahead is zero-width.

**Comment by n0tHappy** [comment]
Thanks @protist
 `let s:default_syntax.char_eqin = '\✹"'`  actually it does not has the `"`,that my fault.
And you can use `$_`, like
```
let s:default_syntax.rxEqIn = '\$_[^\$_`]\+\$_'
let s:default_syntax.char_eqin = '\$_'
```
it works fine. and I don't know why `$m` didn't work.
I tried to config in .vimrc also, but failed too.
`let s:default_syntax = g:vimwiki_syntax_variables['default']`
and even `g:vimwiki_syntax_variables['default']` can not be configed.
It looks like developers didn't provide these configurations.

**Comment by ratfactor** [comment]
That `$` syntax is one of my biggest VimWiki irritations as well.  I wonder if the right answer would be to make a wiki-local option?  Perhaps something like `use_inline_math`, enabled by default, of course.

Then to turn off The Curse of the $ Double Dollars $ for the wiki at `~/my_wiki` you could do this:
```
let g:vimwiki_list = [{'path': '~/my_wiki/', 'use_inline_path': '0'}]
```

If this seems like a good solution, maybe I could get started on it after I'm done with #420.

**Comment by protist** [comment]
> And you can use `$_`, like
> 
> ```
> let s:default_syntax.rxEqIn = '\$_[^\$_`]\+\$_'
> let s:default_syntax.char_eqin = '\$_'
> ```

Huh, I actually thought this should fail, because the `[...]` part is just a character class, i.e. it breaks the regex on the `$` in the middle of `$_foo$bar$_`. However, it seems to work, kinda. Maybe because of the second line. (I haven't delved into the code.)

However, there's a minor problem; instead of hiding the full `$_` when you are not on the line, it only hides the `_` part.

**Comment by n0tHappy** [comment]
hi @protist , on my side, it doesn't have that problem. 
![Untitled](https://user-images.githubusercontent.com/31355958/65096536-e42fd400-d9f7-11e9-8d6f-d22b93450c30.gif)

**Comment by protist** [comment]
@n0tHappy Hm… That does seem to work for me too. I've deleted the test that failed, so I can't replicate it. FWIW `$_foo$bar$_` also fails, and is displayed as italics/emphasis between the `_`.

Anyway, no big deal. I don't think I've ever used the math highlights, so just removing them is fine! Thanks again for the tip. (I'd love to see one of these alternatives in the vanilla code though.)

**Comment by n0tHappy** [comment]
@protist hi, friend, you can try this one:
```
let s:default_syntax.rxEqIn = '\$_[^`]\+\$_'
```
this regex can match `$_foo$bar$_`

**Comment by protist** [comment]
That works, but the `[^$]` part is because the regex is greedy. So this will fail with multiple tags on a single line, e.g. `$_foo$_ and $_bar$_`, where it will highlight both. The following works instead, although it also matches zero-width expressions (between the `$_`), unlike `\+`.

    let s:default_syntax.rxEqIn = '\$_[^`]\{-}\$_'

**Comment by ranebrown** [comment]
Might be worth adding `\[ \]` as an option for math since that would probably have less conflicts (in addition to the disable option).

**Comment by tinmarino** [comment]
Closing <= `\$` is working

**Comment by protist** [comment]
> Closing <= `\$` is working

@tinmarino  I just tested the `dev` branch, and `\$` doesn't seem to work for me.

**Comment by tinmarino** [comment]
@protist  you are right, it only works for markdown, let me fix this for default vimwiki syntax.

**Comment by tinmarino** [comment]
Last commit tends to fix it.

**Comment by protist** [comment]
Thanks @tinmarino.

For me markdown is still broken. If I type in `\$1 \$2`, vim shows `\1 \2`.

For vimwiki syntax, vim displays it fine as `\$1 \$2`, but when I convert to html it renders as `\\(1 \\)2`.

**Comment by tinmarino** [comment]
Thanks @protist for these infos. I only changed the definition of the syntax region for Math but it diregarded the highlight of those symbols (delimiter) and the corresponding convention in default. In other words, I implemented only 1/2 of the feature.
I'll reopen and fix what you spot.

**Comment by protist** [comment]
Thanks @tinmarino. While you are working on this, I might just reiterate that IMO `$` is too common to use for math highlights. I think it should be something less common, like `$$`, etc. FWIW this (and a few other markdown issues) has been annoying me enough that I'm in the process of investigating other Markdown syntax plugins for vim.

**Comment by tinmarino** [comment]
I agree with that other remark @protist as the `$` starting a region is
annoying me too, it is much used in code snippet ...
I'll consider that too, especially if it is not markdown syntax. If it is,
we may have to stick with the `\$` to ensure valid markdown.

**Comment by protist** [comment]
@tinmarino Thanks again. Yes, for me, it's annoying for things like `The apple cost $1, but now it's $2`.

Obviously it's pretty hard to decide what to use because there is no "standard" markdown for math. BTW I found [this](https://github.com/cben/mathdown/wiki/math-in-markdown) interesting survey of syntax if you are interested.

**Comment by brennen** [comment]
> While you are working on this, I might just reiterate that IMO $ is too common to use for math highlights. I think it should be something less common, like $$, etc.

I don't love it either, but keep in mind that this has been the established syntax for quite a while now, and presumably it's in use. We've got [a specification](https://github.com/vimwiki/vimwiki/blob/dev/doc/specification.wiki) for the markup at this point, and I think we should probably start being a bit more deliberate about changes to the syntax, and try to keep the spec up to date and formally versioned.

**Comment by protist** [comment]
>keep in mind that this has been the established syntax for quite a while now

@brennen Good point. I imagine changes would break things for quite a few people.

**Comment by protist** [comment]
Having said that, perhaps an option might be a good compromise, but more work obviously.

**Comment by mcastorina** [comment]
Just ran into this exact issue today (on the `dev` branch). I like the idea of not highlighting `$1 to $2` as that seems like the most common case where a user (myself included) wouldn't want highlighting.

My current workaround is using backticks to render it as code:

```
`$1` to `$2`
```

**Comment by ryspace** [comment]
Ran into this, as I started using vimwiki today for the first time...

> Just ran into this exact issue today (on the `dev` branch). I like the idea of not highlighting `$1 to $2` as that seems like the most common case where a user (myself included) wouldn't want highlighting.
> 
> My current workaround is using backticks to render it as code:
> 
> ```
> `$1` to `$2`
> ```

This works in vim itself, but not when rendered in HTML

After 7 years, is there still not a working way to render this without having to actually modify the source code of the whole plugin?

**Comment by tinmarino** [comment]
Thank @protist for the [comment with pandoc specification](https://github.com/vimwiki/vimwiki/issues/150#issuecomment-864484519).

From that link:

```
Anything between two $ characters will be treated as TeX math. The opening $ must have a non-space character immediately to its right, while the closing $ must have a non-space character immediately to its left, and must not be followed immediately by a digit. Thus, $20,000 and $30,000 won’t parse as math. If for some reason you need to enclose text in literal $ characters, backslash-escape them and they won’t be treated as math delimiters.
```

Thus I created those test cases in [test/issue_150_inline_math.vader](https://github.com/tinmarino/vimwiki/blob/f55e70082856cdd7b32fec555fac3fae21f14c31/test/issue_150_inline_math.vader)

```markdown
12345678901234567890                   # 01 Just a counter
This $math \sum{2, 3}$                 # 02 Math no trap
$This math 3 \times 4 = 12$            # 03 Math all line
This $3 and $4 is not math             # 04 Not math as the last $ is followed by 4
This 3$ and 4$ is not math             # 05 Not math as the first $ is folowed by space
This \$3 and $4 is not math at all     # 06 Not math as first escaped
This $3 and \$4 is not math at all     # 07 Not math as last escaped
This \$3 and \$4 is not math at all    # 08 Not math as both escaped
This $is not math either $320          # 09 Not math as last followd by number (bis)
```

And consider the issue fixed since f55e700828 already in the dev branch. 8 years, that was a long one but some refactoring was needed before.

Thank you all for your contribution, specifically the very constructive comments in this (hard and long) issue.

 🎉  🎉  🎉  🎉  🎉

**Comment by kevxyz** [comment]
What's the workaround today? I used to have this working, but have recently switched from Vundle to native vim plugins, and can't remove the $ syntax. I frequently type two $'s in one line and just want to remove the syntax. What can I do?
